services/device-registry-file/src/main/java/org/eclipse/hono/deviceregistry/file/FileBasedRegistrationService.java

```diff
@@ -520,6 +527,120 @@ private JsonObject convertDevice(final String deviceId, final Device payload) {
         return identities.computeIfAbsent(tenantId, id -> new ConcurrentHashMap<>());
     }
 
+    /**
+     * Return all devices for a tenant.
+     *
+     * The order of devices cannot be guaranteed by the underlying in-memory storage
+     * so the paging may be inconsistent.
+     */
+    @Override
+    public Future<OperationResult<SearchDevicesResult>> searchDevices(final String tenantId, final int pageSize, final int pageOffset,
+                                                                       final List<Filter> filters, final List<Sort> sortOptions, final Span span) {
+
+        final Set allDevicesIds = getDevicesForTenant(tenantId).keySet();
+        List<DeviceWithId> devicesList = new ArrayList<>();
+
+        for (int i = 0; i < allDevicesIds.size(); i++) {
+            final String id = (String) allDevicesIds.toArray()[i];
+            final Device device = getDevicesForTenant(tenantId).get(id).getValue();
+            final DeviceWithId withId = DeviceWithId.from(id, device);
+
+            devicesList.add(withId);
+        }
+
+        final List<Predicate<DeviceWithId>> predicates = new ArrayList<>();
+        for (Filter f : filters) {
+            predicates.add(buildJsonBasedPredicate(f));
+        }
+
+        //Regroup predicates in a single one with AND.
+        final Predicate<DeviceWithId> predicate = predicates.stream()
+                .reduce(x -> true, Predicate::and);
+
+        // 1. apply the filter options to the list
+        devicesList = devicesList
+                .stream()
+                .filter(predicate)
+                .collect(Collectors.toList());
+
+        // 2. sort the selected elements
+        for (Sort s : sortOptions) {
+            devicesList = sortDeviceList(devicesList, s);
+        }
+
+        // 3. limit the results according to the page settings.
+        final List<DeviceWithId> returnDevicesList = new ArrayList<>();
+
+        for (int i = pageOffset; i < devicesList.size() && returnDevicesList.size() < pageSize; i++) {
```

Comment:
pageOffset corresponds to pages and not the place holder of the result in resultset. I think here it should be int `i = pageOffset*PageSize+1`. Eg. When the pageSize is set to 10 and pageOffset to 2, the result should start with 21st device in the resultset and not 2nd.

Reply:
IMO the offset should not be tied to the page size, they are different things. Otherwise If I want to ignore the first 5 results for whatever reason and have a page size of 10, how would I do ? 
Writing this makes me realise that I should not have followed your variables names, as `offset` was more suited for this variable.

Reply:
Yes, the pageOffset contains the number of the page to retrieve. in conjunction with the pageSize (number of entries per page) you can page through the whole result set. If you want to ignore the first entries on a page, then simply retrieve the corresponding page and skip the first entries :-) This is how all APIs that I have stumbled across so far work as well ...

Reply:
Ok I aligned to that.